src/main/scala/org/allenai/dqa/matching/MatchingModel.scala

```diff
@@ -12,68 +12,159 @@ import edu.cmu.dynet.dynet_swig._
 import org.allenai.pnp.ExecutionScore
 import com.google.common.base.Preconditions
 import org.allenai.pnp.Env
+import org.allenai.pnp.CompGraph
 
 /**
  * Structured prediction model for diagram part matching.
  * Given two diagrams with parts, this model predicts a
  * distribution over matchings between the two sets of
  * parts. 
  */
-class MatchingModel(val featureDim: Int, val sourceTargetParam: Parameter,
-    val model: PnpModel) {
+class MatchingModel(val featureDim: Int, val matchIndependent: Boolean,
+    val binaryFactors: Boolean, val model: PnpModel) {
+
+  import MatchingModel._
   
   def apply(source: Diagram, target: Diagram): Pnp[MatchingLabel] = {
     val sourceParts = source.parts
     val targetParts = target.parts
 
     for {
       cg <- computationGraph()
-      preprocessing = preprocess(source, target, cg.cg)
-      matching <- matchRemaining(targetParts.toList, sourceParts.toSet, preprocessing)
+      preprocessing = preprocess(source, target, cg)
+      matching <- matchRemaining(targetParts.toList, sourceParts.toSet, List(), preprocessing)
     } yield {
       MatchingLabel(matching.map(x => (x._1.ind, x._2.ind)).toMap)
     }
   }
 
-  def preprocess(source: Diagram, target: Diagram, cg: ComputationGraph): MatchingPreprocessing = {
+  def preprocess(source: Diagram, target: Diagram, computationGraph: CompGraph): MatchingPreprocessing = {
+    val cg = computationGraph.cg
     val sourceFeatures = source.features.getFeatureMatrix(source.parts, cg)
     val targetFeatures = target.features.getFeatureMatrix(target.parts, cg)
 
-    val weights = parameter(cg, sourceTargetParam)
+    val distanceWeights = parameter(cg, computationGraph.getParameter(DISTANCE_WEIGHTS))
+    val mlpL1W = parameter(cg, computationGraph.getParameter(MLP_L1_W))
+    val mlpL1B = parameter(cg, computationGraph.getParameter(MLP_L1_B))
+    val mlpL2W = parameter(cg, computationGraph.getParameter(MLP_L2_W))
     
     val matchScores = for {
       sourceFeature <- sourceFeatures
     } yield {
       for {
         targetFeature <- targetFeatures
       } yield {
-        // TODO: different scoring function
-        transpose(sourceFeature) * weights * targetFeature
+        // Learn a distance metric
```

Comment:
I find this nested for-yield construct difficult to read, it might be clearer if you pulled out that distance calculation into a function and did something like 

```
for { 
  sourceFeature <- sourceFeatures
} yield allDistances(sourceFeature, targetFeatures)
```

or something like that. I won't insist on it, but it's worth considering.